File: android/app/src/main/java/jaclib/memory/heap/NativeHeapBuffer.java
package jaclib.memory.heap;

import jaclib.memory.Buffer;
import jaclib.memory.Source;

public final class NativeHeapBuffer implements Buffer, Source {

    public boolean d = true;

    public final int b;

    public final NativeHeap a;

    public final int c;

    public NativeHeapBuffer(NativeHeap arg0, int arg1, int arg2) {
        this.b = arg2;
        this.a = arg0;
        this.c = arg1;
    }

    @Override
    public int getSize() {
        return this.b;
    }

    @Override
    public synchronized void a(byte[] arg0, int arg1, int arg2, int arg3) {
        if (arg2 < 0 | arg0 == null | !this.a() | arg1 < 0 | arg1 + arg3 > arg0.length | this.b < arg2 + arg3) {
            throw new RuntimeException();
        }
        this.a.put(this.c, arg0, arg1, arg2, arg3);
    }

    @Override
    public long getAddress() {
        return this.a.getBufferAddress(this.c);
    }

    public synchronized boolean a() {
        return this.a.a() && this.d;
    }

    public synchronized void b() {
        if (this.a()) {
            this.a.deallocateBuffer(this.c);
        }
        this.d = false;
    }

    @Override
    public synchronized void finalize() throws Throwable {
        super.finalize();
        this.b();
    }
}

File: src/jvmMain/kotlin/jaclib/memory/heap/NativeHeapBuffer.kt
package jaclib.memory.heap

import jaclib.memory.Buffer
import jaclib.memory.Source

class NativeHeapBuffer(private val a: NativeHeap, private val d: Int, @JvmField val b: Int) : Buffer, Source {
    private var c = true

    override fun getAddress(): Long {
        return this.a.getBufferAddress(this.d)
    }

    @Synchronized
    private fun a(): Boolean {
        return this.a.a() && this.c
    }

    @Synchronized
    private fun b() {
        if (this.a()) {
            this.a.deallocateBuffer(this.d)
        }
        this.c = false
    }

    @Synchronized
    override fun a(arg0: ByteArray?, arg1: Int, arg2: Int, arg3: Int) {
        if ((arg1 < 0) or (arg0 == null) or !this.a() or (arg3 + arg1 > arg0!!.size) or (arg2 < 0) or (this.b < arg2 + arg3)) {
            throw RuntimeException()
        }
        this.a.put(this.d, arg0, arg1, arg2, arg3)
    }

    @Synchronized
    @Throws(Throwable::class)
    fun finalize() {
//        super.finalize()
        this.b()
    }

    override fun getSize(): Int {
        return this.b
    }
}
